sympy/solvers/solveset.py

```diff
@@ -240,7 +240,7 @@ def _invert_real(f, g_ys, symbol):
 
         if not expo_has_sym:
             res = imageset(Lambda(n, real_root(n, expo)), g_ys)
-            if expo.is_rational:
+            if expo.is_rational or expo.is_Float:
```

Comment:
How does the irrational solution differ from the solution for when expo is a Float?

Reply:
When the exponent is irrational, the the solution is constrained to real numbers greater than zero. The proof of this is presented in issue #21236.

Reply:
have you resolved
- [ ] https://github.com/sympy/sympy/issues/21236#issuecomment-819068154

Reply:
I think it makes more sense to treat the Float as irrational.

Reply:
@oscarbenjamin, moving the Float to the irrational case leads to an EmptySet:
```python
>>> solveset(x**1.0 + 1, x, S.Reals)
EmptySet
>>> solveset(x**1 + 1, x, S.Reals)
FiniteSet(-1)
>>> 1.0 == 1
True
```
@smichr, yes, this [comment](https://github.com/sympy/sympy/issues/21236#issuecomment-819068154) is resolved with my PR:
```python
>>> solveset(sin(x)**1, x, S.Reals)
Union(ImageSet(Lambda(_n, 2*_n*pi + pi), Integers), ImageSet(Lambda(_n, 2*_n*pi), Integers))
>>> solveset(sin(x)*1.0, x, S.Reals)
Union(ImageSet(Lambda(_n, 2*_n*pi + pi), Integers), ImageSet(Lambda(_n, 2*_n*pi), Integers))
```

Reply:
The question is what does `invert` do rather than what does `solveset` do.

Reply:
Also I'm not sure that we want `x**1.0 + 1` to give `-1`. We have to think of a float as being like a range of values and most of those values are irrational.

Reply:
I'm really not sure what's the best way to handle Float here...

Perhaps if there is a possibility that the base is negative and if the exponent is a Float then `solveset` should ultimately return a `ConditionSet`.

Reply:
Maybe this is fine although I do feel uneasy that floats in the input have disappeared from the output.

Reply:
I agree with the unease on converting the float input to integer at the output and that problem lies elsewhere.

Reply:
Does it lie elsewhere? What is the difference in the output of invert for these cases?

Reply:
```python
print(solveset(x**Symbol('rational',rational=1)-2,x,Reals))
Intersection(FiniteSet(2**(1/rational)), Reals)

print(solveset(x**Symbol('irrational',irrational=1)-2,x,Reals))
FiniteSet(2**(1/irrational))
```

Also, the check for even numerator or denominator should only be done for Rational input since `x**(Symbol('e',even=1)/2)` should not be treated like it has an even denominator...or numerator.

Reply:
And if I'm not mistaken, all this is empty code or else the results should be put into a Piecewise since the assumptions on the symbol for which a solution is being sought is stripped of all attributes making it impossible to determine the sign of the base:
```python
        if not base_has_sym:
            rhs = g_ys.args[0]
            if base.is_positive:
                return _invert_real(expo,
                    imageset(Lambda(n, log(n, base, evaluate=False)), g_ys), symbol)
            elif base.is_negative:
                from sympy.core.power import integer_log
                s, b = integer_log(rhs, base)
                if b:
                    return _invert_real(expo, FiniteSet(s), symbol)
                else:
                    return _invert_real(expo, S.EmptySet, symbol)
            elif base.is_zero:
                one = Eq(rhs, 1)
                if one == S.true:
                    # special case: 0**x - 1
                    return _invert_real(expo, FiniteSet(0), symbol)
                elif one == S.false:
                    return _invert_real(expo, S.EmptySet, symbol)
```
e.g.
```
 p=Symbol('p',positive=1)
 solveset(p**x-2,p)
ConditionSet(p, Eq(p**x - 2, 0), Complexes)
 solveset(p**x-2,p,S.Reals)
ConditionSet(p, Eq(p**x - 2, 0), Reals)
 solveset(p**x-2,p,(x>0).as_set())
ConditionSet(p, Eq(p**x - 2, 0), Interval.open(0, oo))

```

Reply:
Condition for a denominator of `e` to be even is `e.is_zero is False and (1/e).is_even`; the numerator is even if `e.is_integer and e.is_even`.

Reply:
Yes, it shouldn't be trying to determin the sign of the base. That's why I suggested:
```python
                g_ys_pos = g_ys & Interval(0, oo)
                res = imageset(Lambda(n, real_root(n, expo)), g_ys_pos)
```
Instead this restricts the possible range of the base.